compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -846,12 +846,17 @@ impl<'tcx> OpaqueHiddenType<'tcx> {
         opaque_def_id: LocalDefId,
         tcx: TyCtxt<'tcx>,
     ) -> Result<DiagnosticBuilder<'tcx>, ErrorGuaranteed> {
+
         if let Some(diag) = tcx
             .sess
             .dcx()
             .steal_diagnostic(tcx.def_span(opaque_def_id), StashKey::OpaqueHiddenTypeMismatch)
         {
-            diag.cancel();
+            // We used to cancel here for slightly better error messages, but
+            // cancelling stashed diagnostics is no longer allowed because it
+            // causes problems when tracking whether errors have actually
+            // occurred.
+            diag.emit();
```

Comment:
would we get the same benefits by turning it into a delayed bug?

Reply:
+1 on this.

Reply:
Unfortunately, the subequent commits forbid that. The reason being that it will mess with the error counts. A stashed error gets counted by `DiagCtxt::err_count` and `DiagCtxt::err_count_excluding_lint_erss`. If it could later be downgraded to a delayed bug then that would no longer be counted by those methods, and the error count would go down. That's not as bad as conjuring an `ErrorGuaranteed` out of nothing, but it's still the kind of laxness that I am desperately trying to eliminate from the stashed diagnostics implementation.

In general, the old stashed diagnostics implementation was wildly unsafe and overpowered, and I'm having to introduce some limitations to fix the problems.